### ISSUE TITLE
Intermittent "Common control call failed" error creating (or updating?) DateTimePicker

### ISSUE THREAD

**Issue by daniels220** [dolphinsmalltalk/Dolphin#1045]
I'm seeing a very occasional "Common control call failed" error in `dtmSetSystemTime:lpSysTime:`, when a DateTimePicker's model is freshly assigned, and possibly also when its value changes. The error doesn't give any details, and I don't know enough about Win32, never mind the Common Controls, to have any idea if there are details available somewhere. The most common stacktrace looks something like:

```
DateTimePicker(ControlView)>>errorInCommonControlCall
DateTimePicker>>dtmSetSystemTime:lpSysTime:
...
DateTimePicker(View)>>onModelChanged
DateTimePicker(View)>>model:
DatePresenter(Presenter)>>connectView
DatePresenter(Presenter)>>view:
...
MyShell(Presenter)>>attachSubPresenterViews:
...
MyShell(Presenter)>>createView:
```

Any ideas (a) what the problem is and how to actually fix it, or failing that (b) how to get more info about the error, or failing that (c) any possible workaround (i.e. what would I have to do to retry with as little disruption as possible—is literally just retrying likely to work? Destroy/recreate the DateTimePicker first? Call `pumpMessages` (I hope not, that would be Wrong even if it worked...)?

**Comment by blairmcg** [comment]
Unfortunately the common controls typically don't provide any richer error information than success/failure, and the documentation doesn't specify any reasons for failure either.

It can sometimes be instructive to look at the relevant bit of .net framework source code, in this case in the [Winforms DateTimePicker](https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DateTimePicker.cs,1036). Certainly the .net code is not checking the return value, but there is no other information (like comments) that would clarify whether that is due to deliberate choice or just sloppiness. On line 1020 one can see that the value is checked to see whether it is in range. This is one reason the common control will return failure, e.g.:

```smalltalk
dtp := DateTimePicker show.
dtp range: {DateAndTime now - 10 days. DateAndTime now + 10 days}.
"This will fail with the dreaded 'Common control call failed'"
dtp value: DateAndTime now - 11 days
```

We should change the code to follow the approach taken in WinForms and validate that the value is in range and throw our own informative error if not, rather than relying on the the generic failure from the control.

What I suggest you do is to patch the method to stop throwing an error, and instead log the failure somehow with the values of the parameters. You might also want to log any other non-default settings, such as the min/max range, or if there is a custom format. The control may work correctly anyway. If it doesn't the behaviour may show a clue, or perhaps there might be some pattern in the parameter values or configuration.

**Comment by daniels220** [comment]
AHA, figured it out! Long story short, the problem was a `SYSTEMTIME` struct with `wMilliseconds = 1000`, as a result of `DateAndTime>>asSYSTEMTIME` doing `(secs fractionPart * 1000) rounded`, which produces 1000 about 1/2000th of the time when starting with `DateAndTime now`—explains why it happened so rarely...

The trivial fix is to change that `rounded` to `truncated` (and I notice the same should be done in `Time>>asSYSTEMTIME`), but this is arguably incorrect in that it produces a larger deviation than necessary from the "true" value. However doing better—by rounding the entire [DateAnd]Time to the nearest millisecond *first*, then computing h:m:s.xx—is ugly and comparatively expensive, and in any case the old TimeStamp didn't even bother with the milliseconds to begin with, nor does it matter for the purposes of this control, so...sound like a good solution? I can make a PR for whichever approach (truncate or pre-round to nearest ms) seems best.

**Comment by blairmcg** [comment]
Ah interesting. I think it is better to just truncate it - it isn't quarter past nine at 9:14 and 31 seconds, it is still 9:14.

**Comment by daniels220** [comment]
Well, the question is about what to do with `9:14:59.9995`—should it be `9:15:00.000`, or `9:14:59.999`? The former seems technically more correct, but the difference is so trivial, and so rare, that I *am* ultimately inclined to go with the truncation...

**Comment by blairmcg** [comment]
I don't think it makes any difference what the time is being measured in, until the unit ticks over, it is still the time of the last tick, even if closer to the next tick. And as you say, it is unimportant for the uses of SYSTEMTIME, which is not intended to be a very precise measure.

**Comment by daniels220** [comment]
Alright, will do.

**Comment by blairmcg** [comment]
Fixed by #1046
